File: frontend/src/api/courses.ts
import { apiClient } from './client';
import type { Course, CourseAssignment } from '@/types';

export async function listCourses(
  limit: number = 20,
  offset: number = 0,
): Promise<Course[]> {
  const response = await apiClient.get<Course[]>('/courses', {
    params: { limit, offset },
  });
  return response.data;
}

export async function getCourse(courseId: number): Promise<Course> {
  const response = await apiClient.get<Course>(`/courses/${courseId}`);
  return response.data;
}

export interface CourseInput {
  title: string;
  description?: string | null;
  content?: string | null;
  duration_minutes?: number;
}

export async function createCourse(data: CourseInput): Promise<Course> {
  const response = await apiClient.post<Course>('/courses', {
    title: data.title,
    description: data.description ?? null,
    content: data.content ?? null,
    duration_minutes: data.duration_minutes ?? 0,
  });
  return response.data;
}

export async function updateCourse(courseId: number, data: Partial<CourseInput>): Promise<Course> {
  const response = await apiClient.patch<Course>(`/courses/${courseId}`, data);
  return response.data;
}

export async function publishCourse(courseId: number): Promise<Course> {
  const response = await apiClient.post<Course>(`/courses/${courseId}/publish`);
  return response.data;
}

export async function assignCourse(courseId: number, userIds: string[]): Promise<CourseAssignment[]> {
  const response = await apiClient.post<CourseAssignment[]>(`/courses/${courseId}/assign`, {
    user_ids: userIds,
  });
  return response.data;
}

export async function updateProgress(courseId: number, progress_percent: number): Promise<CourseAssignment> {
  const response = await apiClient.post<CourseAssignment>(`/courses/${courseId}/progress`, null, {
    params: { progress_percent },
  });
  return response.data;
}

export async function myAssignments(): Promise<CourseAssignment[]> {
  const response = await apiClient.get<CourseAssignment[]>('/courses/assignments/me');
  return response.data;
}

File: backend/app/api/v1/endpoints/courses.py
"""Эндпоинты LMS (курсы и назначения)."""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Body, Depends, Query, Request, status

from app.core.dependencies import (
    ActiveSubscriptionContext,
    CurrentContext,
    DbSession,
    require_roles,
)
from app.models import OrgRole
from app.schemas.course import (
    CourseAssignmentResponse,
    CourseCreate,
    CourseResponse,
    CourseUpdate,
)
from app.services.course_service import CourseService

router = APIRouter()


@router.get(
    "",
    response_model=list[CourseResponse],
    summary="Список курсов",
    description="Получение опубликованных курсов организации.",
)
async def list_courses(
    ctx: ActiveSubscriptionContext,
    session: DbSession,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> list[CourseResponse]:
    service = CourseService(session)
    courses = await service.list_courses(ctx.organization_id, limit=limit, offset=offset)
    return [CourseResponse.model_validate(course) for course in courses]


@router.post(
    "",
    response_model=CourseResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Создать курс",
    dependencies=[Depends(require_roles(OrgRole.ORG_OWNER))],
)
async def create_course(
    request: Request,
    data: CourseCreate,
    ctx: CurrentContext,
    session: DbSession,
) -> CourseResponse:
    service = CourseService(session)
    return await service.create_course(
        ctx.organization_id,
        data,
        request_id=getattr(request.state, "request_id", None),
    )


@router.patch(
    "/{course_id}",
    response_model=CourseResponse,
    summary="Обновить курс",
    dependencies=[Depends(require_roles(OrgRole.ORG_OWNER))],
)
async def update_course(
    course_id: int,
    request: Request,
    data: CourseUpdate,
    ctx: CurrentContext,
    session: DbSession,
) -> CourseResponse:
    service = CourseService(session)
    return await service.update_course(
        course_id,
        ctx.organization_id,
        data,
        request_id=getattr(request.state, "request_id", None),
    )


@router.post(
    "/{course_id}/publish",
    response_model=CourseResponse,
    summary="Публиковать курс",
    dependencies=[Depends(require_roles(OrgRole.ORG_OWNER))],
)
async def publish_course(
    course_id: int,
    request: Request,
    ctx: CurrentContext,
    session: DbSession,
) -> CourseResponse:
    service = CourseService(session)
    return await service.publish_course(
        course_id,
        ctx.organization_id,
        request_id=getattr(request.state, "request_id", None),
    )


@router.post(
    "/{course_id}/assign",
    response_model=list[CourseAssignmentResponse],
    summary="Назначить курс пользователям",
    dependencies=[Depends(require_roles(OrgRole.ORG_OWNER))],
)
async def assign_course(
    course_id: int,
    request: Request,
    ctx: CurrentContext,
    session: DbSession,
    user_ids: Annotated[list[UUID], Body(embed=True, description="Список пользователей для назначения")],
) -> list[CourseAssignmentResponse]:
    service = CourseService(session)
    return await service.assign_course(
        course_id,
        ctx.organization_id,
        user_ids,
        actor_id=ctx.user.id,
        request_id=getattr(request.state, "request_id", None),
    )


@router.post(
    "/{course_id}/progress",
    response_model=CourseAssignmentResponse,
    summary="Обновить прогресс курса для текущего пользователя",
)
async def update_progress(
    course_id: int,
    request: Request,
    ctx: ActiveSubscriptionContext,
    session: DbSession,
    progress_percent: Annotated[int, Query(ge=0, le=100)],
) -> CourseAssignmentResponse:
    service = CourseService(session)
    return await service.update_progress(
        course_id=course_id,
        organization_id=ctx.organization_id,
        user_id=ctx.user.id,
        progress_percent=progress_percent,
        request_id=getattr(request.state, "request_id", None) if request else None,
    )


@router.get(
    "/assignments/me",
    response_model=list[CourseAssignmentResponse],
    summary="Назначенные мне курсы",
)
async def my_assignments(
    ctx: ActiveSubscriptionContext,
    session: DbSession,
) -> list[CourseAssignmentResponse]:
    service = CourseService(session)
    return await service.list_assignments_for_user(
        user_id=ctx.user.id,
        organization_id=ctx.organization_id,
    )
